### PR TITLE
Adding ability to perturb a vector by some factor

### LINKED ISSUE #260: Feature: Add ability to perturb a vector by some factor
For various purposes, it is often useful to perturb a vector by some random amount scaled by some factor. 

The following function takes in a list of vectors and some factor `eps` in which each vector in the list is randomly perturbed:

```py
import numpy as np


def perturb_vectors(vectors: list[np.ndarray], eps: float = 0.1) -> list[np.ndarray]:
    """Perturb the vectors by adding a small random number to each element.

    :param vectors: List of vectors to perturb.
    :param eps: Amount by which to perturb vectors.
    :return: Resulting list of perturbed vectors by a factor of epsilon.
    """
    perturbed_vectors: list[np.ndarray] = []
    for i, v in enumerate(vectors):
        perturbed_vectors.append(v + np.random.randn(v.shape[0]) * eps)

        # Normalize the vectors after perturbing them.
        perturbed_vectors[i] = perturbed_vectors[i] / np.linalg.norm(perturbed_vectors[i])
    return np.array(perturbed_vectors)
```

The function can probably be written in a more concise manner and also requires documentation and tests.

### PR DESCRIPTION
**Pull request by Shivansh20128**
## Description
Provide a brief description of the PR's purpose here. If your PR is supposed to fix an existing issue, use
a [keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link your PR to the issue.

Adding the ability to perturb a vector by some random value by adding a submodule in _matrix_ops_.
This PR includes the docstring, the function as described in the issue mentioned, and the tests for the feature function.
#Resolves #260
## Changes
Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below.

  -  [x] Added the function `perturb_vectors`
  -  [x] Added docstrings for the function `perturb_vectors`
  -  [x] Added tests for the function `perturb_vectors` in file `test_perturb_vector.py`

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/matrix_ops/perturb_vectors.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+"""
+
+import numpy as np
+
+
+def perturb_vectors(vectors: list[np.ndarray], eps: float = 0.1) -> list[np.ndarray]:
+    """Perturb the vectors by adding a small random number to each element.
+
+    :param vectors: List of vectors to perturb.
+    :param eps: Amount by which to perturb vectors.
```
Comment (purva-thakre) [inline toqito/matrix_ops/perturb_vectors.py:13]: this function is also missing an example

Comment (Shivansh20128) [reply toqito/matrix_ops/perturb_vectors.py:13]: Thank You for the review. I have added an example now.

File: toqito/matrix_ops/perturb_vectors.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    :return: Resulting list of perturbed vectors by a factor of epsilon.
+
+    Example:
+        >>> from toqito.matrix_ops import perturb_vectors
+        >>> import numpy as np
+        >>> vectors = [np.array([1.0, 2.0]), np.array([3.0, 4.0])]
+        >>> perturbed_vectors = perturb_vectors(vectors, eps=0.1)
```
Comment (purva-thakre) [inline toqito/matrix_ops/perturb_vectors.py:20]:
```suggestion
        >>> perturb_vectors(vectors, eps=0.1)
```
Show the output as doctest verifies this.  You can use `# doctest: +SKIP` for doctest failures as we expect the output to be different due to `randn`.

https://github.com/vprusso/toqito/blob/4a4ce5e3e65b74348fcc1abe106b7ccf443bc050/toqito/rand/random_unitary.py#L24

File: toqito/matrix_ops/tests/test_perturb_vector.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    [
+        ([np.array([1, 0, 0]), np.array([0, 1, 0])]),
+        ([np.array([1, 1, 1])]),
+    ],
+)
+def test_zero_perturbation(vectors):
+    """Test that if eps = 0, the vectors remain the same."""
+    perturbed_vectors = perturb_vectors(vectors, eps=0.0)
+    for i in range(len(vectors)):
+        np.testing.assert_array_almost_equal(vectors[i], perturbed_vectors[i])
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_perturb_vector.py:66]: lets not use np.testing. We are trying to move away from it #488

File: toqito/matrix_ops/perturb_vectors.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def perturb_vectors(vectors: list[np.ndarray], eps: float = 0.1) -> list[np.ndarray]:
+    """Perturb the vectors by adding a small random number to each element.
+
+    :param vectors: List of vectors to perturb.
+    :param eps: Amount by which to perturb vectors.
+    :return: Resulting list of perturbed vectors by a factor of epsilon.
+
+    Example:
```
Comment (purva-thakre) [inline toqito/matrix_ops/perturb_vectors.py:16]: Replace with

```
    Example
    ==========
```

Comment (Shivansh20128) [reply toqito/matrix_ops/perturb_vectors.py:16]: done!

File: toqito/matrix_ops/tests/test_perturb_vector.py (hunk 3 above)
Comment (Shivansh20128) [reply toqito/matrix_ops/tests/test_perturb_vector.py]: done! I hope its fine now.

File: toqito/matrix_ops/perturb_vectors.py (hunk 2 above)
Comment (Shivansh20128) [reply toqito/matrix_ops/perturb_vectors.py]: done!

**Comment by vprusso** [comment]
Thanks for the changes and updates, @Shivansh20128 ! LGTM, but I'll ping @purva-thakre to see if she has any additional final comments. Great work, as always!

File: toqito/matrix_ops/perturb_vectors.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    ==========
+
+        >>> from toqito.matrix_ops import perturb_vectors
+        >>> import numpy as np
+        >>> vectors = [np.array([1.0, 2.0]), np.array([3.0, 4.0])]
+        >>> perturb_vectors(vectors, eps=0.1) # doctest: +SKIP
+
```
Comment (purva-thakre) [inline toqito/matrix_ops/perturb_vectors.py:23]: @Shivansh20128 You have to add the output manually here. Right now, the RTD output is blank for the example.

image



See below for example:

https://github.com/vprusso/toqito/blob/9b355a57c98c83e4d6342f28d1b396b27bcf8abb/toqito/rand/random_povm.py#L21-L34

image

File: toqito/matrix_ops/perturb_vectors.py (hunk 2 above)
Comment (purva-thakre) [reply toqito/matrix_ops/perturb_vectors.py]: This is not done yet. See https://github.com/vprusso/toqito/pull/861#discussion_r1826984020

File: toqito/matrix_ops/perturb_vectors.py (hunk 5 above)
Comment (Shivansh20128) [reply toqito/matrix_ops/perturb_vectors.py:25]: Added! I hope its fine now.

**Review by purva-thakre — APPROVED** [review]
LGTM. Thank you for your patience @Shivansh20128 !

**Pull request merged**